p2p/gater.go

```diff
@@ -0,0 +1,44 @@
+package p2p
+
+import (
+	"github.com/libp2p/go-libp2p-core/connmgr"
+	"github.com/libp2p/go-libp2p-core/control"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	libp2p_dht "github.com/libp2p/go-libp2p-kad-dht"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+type Gater struct{}
+
+func NewGater(disablePrivateIPScan bool) connmgr.ConnectionGater {
+	if !disablePrivateIPScan {
+		return nil
```

Comment:
Returning nil from constructor seems like bad design for me. Maybe exactly here it's ok, but such style leads to runtime panics.